### ISSUE TITLE
(Staggered) mixed precision inverter is broken with reconstruction 9

### ISSUE THREAD

**Issue by mathiaswagner** [lattice/quda#333]
I tried using the staggered mixed precision inverter in MILC with reconstruction and saw convergence issues. I can reproduce the issue with staggered_invert_test on a single node.

For reference pure double:

```
test_invert_double18_double18.txt:CG: Convergence at 1825 iterations, L2 relative residual: iterated = 9.865356e-08, true = 9.865356e-08
```

Double single without reconstruction and pure double with 9 reconstruct:

```
test_invert_double18_single18.txt:CG: Convergence at 1829 iterations, L2 relative residual: iterated = 9.975623e-08, true = 9.975623e-08
test_invert_double9_double9.txt:CG: Convergence at 1832 iterations, L2 relative residual: iterated = 9.997816e-08, true = 9.997816e-08
```

And finally mixed precision with reconstruction:

```
test_invert_double9_single18.txt:CG: Convergence at 715 iterations, L2 relative residual: iterated = 1.437793e+01, true = 1.437793e+01
test_invert_double9_single9.txt:CG: Convergence at 717 iterations, L2 relative residual: iterated = 1.443085e+01, true = 1.443085e+01
```

The name of the output is `test_invert_<prec><recon>_<prec_sloppy><recon_sloppy>.txt`

There are more combinations in this parameter space but something is going wrong when combining reconstruction with mixed precision.

**Comment by mikeaclark** [comment]
Is there a way in MILC you can set the boundary conditions to be all periodic?  I wonder if it's related to boundary conditions getting screwed up.

**Comment by mikeaclark** [comment]
This is the develop branch isn't it?

**Comment by mathiaswagner** [comment]
This is the branch for the milc baseline (essentially develop + NVTX).
The results I copied above are from the internal QUDA tests:
```
./staggered_invert_test --recon 9 --recon_sloppy 9 --prec double --prec_sloppy single
```

**Comment by mathiaswagner** [comment]
Adding more datapoints:
```
test_invert_double13_single13.txt:CG: Convergence at 1841 iterations, L2 relative residual: iterated = 9.950236e-08, true = 9.950236e-08
test_invert_double13_single18.txt:CG: Convergence at 1878 iterations, L2 relative residual: iterated = 9.979013e-08, true = 9.979013e-08
```

13 reconstruct seems to be fine.

**Comment by mikeaclark** [comment]
Ah sorry, I understand now.  For some reason I inserted a "not" in relation to whether you could reproduce the issue.

Can you switch off the boundary condition then to see if this affects this?

**Comment by mathiaswagner** [comment]
I still have not turned off the boundary conditions but generated a more complete matrix of the impact of reconstruction on the staggered solver:



## Single precision and reconstruction

### Baseline: Pure Single

```
single18_single18.txt:CG:  1739 iterations, L2 : iterated = 1.512053e-05, true = 1.512053e-05
```


### Reconstruction
```
! single13_single13.txt:CG:  1842 iterations, L2 : iterated = 1.508971e-05, true = 1.508971e-05
! single9_single9.txt:CG:  1933 iterations, L2 : iterated = 1.505647e-05, true = 1.505647e-05
```


## Double precision and reconstruction

### Baseline
```
double18_double18.txt:CG:  1825 iterations, L2 : iterated = 9.865356e-08, true = 9.865356e-08
```

### Reconstruction
```
double13_double13.txt:CG:  1833 iterations, L2 : iterated = 9.952464e-08, true = 9.952464e-08
double9_double9.txt:CG:  1832 iterations, L2 : iterated = 9.997816e-08, true = 9.997816e-08
```

## Mixed precision (Double single)

### Baseline
```
double18_single18.txt:CG:  1829 iterations, L2 : iterated = 9.975623e-08, true = 9.975623e-08
```

### Reconstrction (identical for both precisions)
```
double13_single13.txt:CG:  1841 iterations, L2 : iterated = 9.950236e-08, true = 9.950236e-08
X double9_single9.txt:CG:  717 iterations, L2 : iterated = 1.443085e+01, true = 1.443085e+01
```

### Reconstruction (only for sloppy precision)
```
double18_single13.txt:CG:  1829 iterations, L2 : iterated = 9.988407e-08, true = 9.988407e-08
double18_single9.txt:CG:  1829 iterations, L2 : iterated = 9.977784e-08, true = 9.977784e-08
```

### Reconstruction (only for high precision)
```
! double13_single18.txt:CG:  1878 iterations, L2 : iterated = 9.979013e-08, true = 9.979013e-08
X double9_single18.txt:CG:  715 iterations, L2 : iterated = 1.437793e+01, true = 1.437793e+01
```

## Summary ###
* When using double precision and 9 reconstruct for the high precision I never saw convergence for mixed precision.
* For pure single precision reconstruction increases the iteration count. That might eat up the improvements from the reconstruction and finally increase the time to solution. (Needs to be checked)
* For pure double precision reconstruction seems to be fine. It only slightly increases the iteration count (here < 1%)
* For mixed precision using reconstruction only on the sloppy precision links is perfectly fine. Probably the high-precision update fixes all possible errors that come from reconstruction in the sloppy precision. (In other words: It might be that the sloppy precision is just slightly more sloppy with reconstruction)
* For mixed precision and using reconstruction on the high precision links: Broken for using 9 reconstruct on the high precision links and iteration count increase by more than 1% for using double13.

**Comment by mikeaclark** [comment]
Thanks for the summary.  It appears this behavior is consistent with the copy gauge from double9 to single/half (any reconstruct).  Is that correct?

I wouldn't read anything into the increase in iteration count for pure single precision with reconstruction.  Given that it fails to converge anyway, the increase in iteration count isn't significant I believe.

**Comment by mathiaswagner** [comment]
Yes, it looks like copying from double9 to anything is strange. Copying from double13 to single18 also is unexpected as it has a higher iteration count than double13,single13.
Copying from single9 to single13,18 is also affected. But I don't know why anyone would run with high precision single9 and sloppy precision single13 or single 18. I had it in my test matrix for completeness but excluded it from the summary.

For the increased iteration count: Not sure about that. The time to solution is:
```
test_invert_single18_single18.txt:done: total time = 6.33 secs, compute time = 6.31115 secs, 1739 iter / 6.31115 secs = 110.682 gflops, 
test_invert_single13_single13.txt:done: total time = 5.91 secs, compute time = 5.88962 secs, 1842 iter / 5.88962 secs = 125.829 gflops, 
test_invert_single9_single9.txt:done: total time = 5.74 secs, compute time = 5.72255 secs, 1933 iter / 5.72255 secs = 136.103 gflops

```

I had not expected such an increase in iteration. It seems that reconstruction introduced an error. We can treat it like mixed precision though:
```
test_invert_single18_single9.txt:CG: Convergence at 1762 iterations, L2 relative residual: iterated = 1.510149e-05, true = 1.510149e-05
test_invert_single18_single9.txt:done: total time = 5.22 secs, compute time = 5.20572 secs, 1762 iter / 5.20572 secs = 136.019 flops
```

And use single9 as sloppy precision and single18 as high precision. That is the best time-to-solution.

**Comment by mathiaswagner** [comment]
Same arguments for the error also apply for mixed precision solvers:
```
test_invert_double18_double18.txt:done: total time = 12.94 secs, compute time = 12.9091 secs, 1825 iter / 12.9091 secs = 56.5541 gflops,
test_invert_double18_single18.txt:done: total time = 6.75 secs, compute time = 6.72528 secs, 1829 iter / 6.72528 secs = 108.792 gflops, 
test_invert_double18_single13.txt:done: total time = 5.98 secs, compute time = 5.93397 secs, 1829 iter / 5.93397 secs = 123.299 gflops, 
test_invert_double18_single9.txt:done: total time = 5.54 secs, compute time = 5.49486 secs, 1829 iter / 5.49486 secs = 133.153 gflops, 
```

We can play it safe and never reconstruct the high precision. It should have a negligible performance impact. double13, single9 is slightly slower (5.56 vs 5.43) than double18, single9 as the iteration count (1841 / 1829) is increased. No timings for double9, single9 as that does not work.

**Comment by mathiaswagner** [comment]
For pure double precision double18, double13 is best. double9 is probably slower due to increased register usage. double13, double13 is slower than double18, double13 as iteration count increases.

**Comment by mathiaswagner** [comment]
Rechecked after #489:

```
./tests/staggered_invert_test --recon 18 --recon-sloppy 9 --prec double --prec-sloppy single --niter 2000
```
and
```
./tests/staggered_invert_test --recon 9 --recon-sloppy 9 --prec double --prec-sloppy double --niter 2000
```

are fine.

```
./tests/staggered_invert_test --recon 9 --recon-sloppy 9 --prec double --prec-sloppy single --niter 2000
```
still fails with

```
WARNING: CG: new reliable residual norm 8.324691e+03 is greater than previous reliable residual norm 3.406423e+03 (total #inc 2)
WARNING: CG: solver exiting due to too many true residual norm increases
```

**Comment by maddyscientist** [comment]
Closed with #776
